### ISSUE TITLE
Possible memory leak when loading a workflow with a query

### ISSUE THREAD

**Issue by ovargasmahisoft** [temporalio/sdk-go#774]
## Expected Behavior

When querying a workflow, the worker should add the workflow to the internal cache.

## Actual Behavior

When the Worker starts and the first interaction with the workflow is a Query, the workflow is not cached, you can notice this behavior because the events are replayed from scratch every time.

The memory leak occurs because every time the query is handled, a new instance of the workflow is allocated by the worker. It only happens when the workflow stays awaiting new signals.

## Steps to Reproduce the Problem

Consider a workflow like

```go
package workflow

import (
	"go.temporal.io/sdk/workflow"
)

func ReproduceML(ctx workflow.Context, param string) error {
	log := workflow.GetLogger(ctx)
	log.Info("Starting new Workflow")
	defer func() {
		log.Info("Exit from workflow")
	}()
	state := param
	var receivedSignals []string

	err := workflow.SetQueryHandler(ctx, "getState", func() (string, error) {
		return state, nil
	})

	if err != nil {
		log.Error("SetQueryHandler getState failed", "Error", err)
		return err
	}

	channel := workflow.GetSignalChannel(ctx, "ml-signals")
	for {

		if state == "complete" {
			break
		}

		var signal string
		_ = channel.Receive(ctx, &signal)
		// Do something special with the signal and calculate the state
		state = signal

		// Forcing memory growing, so it can be noticed it in the metrics
		receivedSignals = append(receivedSignals, signal)
	}

	log.Info("Workflow finished")
	return nil
}

```

1. Start the worker
2. Create a workflow
3. Signal the workflow around 500 times to generate some data.
4. Restart the worker
5. ONLY query the workflow several times and notice how the memory grows.

Verifying:

- In the logs, you can notice an entry "Starting new Workflow" for every query and never the "Exit from workflow" meaning the workflow stayed allocated.
- If we disable the cache with `worker.SetStickyWorkflowCacheSize(0)`, the logs show "Starting new Workflow" and "Exit from workflow" with every query execution and the memory stops growing.
- With the cache enabled, If after the 4th step, the first interaction is a signal, the "Starting new Workflow" is shown only once, meaning the cache was properly set, and queries to that workflow will be returned faster and the memory won't grow as before.

Final note: with a workflow with a similar structure as the above, the memory increased to 6Gb+, and without modifying code after doing one of the last two bullet points the memory stays below 30 Mb.

## Specifications

- Version:
  go.temporal.io/sdk v1.14.0
  Golang 1.17.8
- Platform:
  Darwin x86_64

**Comment by cretz** [comment]
To confirm, for step #5, it is not reusing the cached workflow for each query call after the first one? If so, this does look like a bug and I will investigate.

**Comment by ovargasmahisoft** [comment]
I put several breakpoints in the code, and I noticed that when the Query is executed just after the worker starts, the [putWorkflowContext](https://github.com/temporalio/sdk-go/blob/v1.14.0/internal/internal_worker_cache.go#L141) function is never called, that's why I believe the cache is not set, but if later you send a signal, the cache is set, and from there the Queries reuse it.

**Comment by ovargasmahisoft** [comment]
I modified a bit your test to show you what I'm seeing

<details>
  <summary>main.go</summary>

```go

package main

import (
	"context"
	"errors"
	"fmt"
	"log"
	"runtime"
	"time"

	"github.com/google/uuid"
	"github.com/uber-go/tally/v4"
	"go.temporal.io/api/serviceerror"
	"go.temporal.io/sdk/client"
	sdktally "go.temporal.io/sdk/contrib/tally"
	"go.temporal.io/sdk/worker"
	"go.temporal.io/sdk/workflow"
)

func main() {
	if err := run(); err != nil {
		log.Fatal(err)
	}
}

func run() error {
	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	log.Printf("Creating client")
	c, err := client.NewClient(client.Options{})
	if err != nil {
		return fmt.Errorf("failed creating client: %w", err)
	}
	defer c.Close()

	log.Printf("Starting worker 1")
	taskQueue := uuid.NewString()
	worker1 := worker.New(c, taskQueue, worker.Options{})
	worker1.RegisterWorkflow(SignalCountingWorkflow)
	if err := worker1.Start(); err != nil {
		return fmt.Errorf("failed starting worker client: %w", err)
	}
	defer func() {
		if worker1 != nil {
			worker1.Stop()
		}
	}()

	log.Printf("Starting workflow")
	run, err := c.ExecuteWorkflow(ctx, client.StartWorkflowOptions{TaskQueue: taskQueue}, SignalCountingWorkflow)
	if err != nil {
		return fmt.Errorf("failed starting workflow: %w", err)
	}

	log.Printf("Sending 500 signals")
	for i := 0; i < 500; i++ {
		if err := c.SignalWorkflow(ctx, run.GetID(), "", "my-signal", "some-arg"); err != nil {
			return fmt.Errorf("failed signalling workflow: %w", err)
		}
	}

	log.Printf("Waiting for query to show 500 signals received")
	var signalCount int
	for start := time.Now(); signalCount != 500 && time.Since(start) < 10*time.Second; {
		time.Sleep(50 * time.Millisecond)
		val, err := c.QueryWorkflow(ctx, run.GetID(), "", "get-signal-count")
		// Ignore query failed because it means query may not be registered yet
		var queryFailed *serviceerror.QueryFailed
		if errors.As(err, &queryFailed) {
			continue
		} else if err != nil {
			return fmt.Errorf("failed querying: %w", err)
		} else if err := val.Get(&signalCount); err != nil {
			return fmt.Errorf("failed decoding query: %w", err)
		}
	}
	if signalCount != 500 {
		return fmt.Errorf("expected signal count never reached")
	}

	log.Printf("Stopping worker")
	worker1.Stop()
	worker1 = nil

	log.Printf("Starting worker 2")
	// Recreate client to capture metrics
	testScope := tally.NewTestScope("", nil)
	c, err = client.NewClient(client.Options{MetricsHandler: sdktally.NewMetricsHandler(testScope)})
	if err != nil {
		return fmt.Errorf("failed creating client: %w", err)
	}
	defer c.Close()
	worker2 := worker.New(c, taskQueue, worker.Options{})
	worker2.RegisterWorkflow(SignalCountingWorkflow)
	if err := worker2.Start(); err != nil {
		return fmt.Errorf("failed starting worker client: %w", err)
	}
	defer worker2.Stop()

	log.Printf("Performing several queries")
	log.Printf("before querying")
	printMemoryUsage()
	// Test 1st with the following line commented, and then uncomment it and test again
	// Check the memory usage
	//
	// _ = c.SignalWorkflow(ctx, run.GetID(), "", "my-signal", "some-arg")
	for i := 0; i < 500; i++ {
		printMemoryUsage()
		if _, err := c.QueryWorkflow(ctx, run.GetID(), "", "get-signal-count"); err != nil {
			return fmt.Errorf("failed querying: %w", err)
		}
		// Get cache size metric
		var cacheSize float64
		for _, gauge := range testScope.Snapshot().Gauges() {
			if gauge.Name() == "temporal_sticky_cache_size" {
				cacheSize += gauge.Value()
			}
		}
		log.Printf("Query #%v complete, cache size: %v", i+1, cacheSize)
	}
	return nil
}

func printMemoryUsage() {
	var m runtime.MemStats
	runtime.ReadMemStats(&m)
	fmt.Printf("Alloc = %v MiB", bToMb(m.Alloc))
	fmt.Printf("\tTotalAlloc = %v MiB", bToMb(m.TotalAlloc))
	fmt.Printf("\tSys = %v MiB", bToMb(m.Sys))
	fmt.Printf("\tNumGC = %v\n", m.NumGC)
}

func SignalCountingWorkflow(ctx workflow.Context) error {
	log.Printf("Workflow starts")
	defer log.Printf("Workflow Exit")
	workflow.GetLogger(ctx).Info("Starting workflow")
	log.Printf("Running workflow code (replaying? %v)", workflow.IsReplaying(ctx))
	defer workflow.GetLogger(ctx).Info("Completing workflow")
	defer func() { log.Printf("Finishing workflow code (replaying? %v)", workflow.IsReplaying(ctx)) }()

	signalCount := 0
	err := workflow.SetQueryHandler(ctx, "get-signal-count", func() (int, error) { return signalCount, nil })
	if err != nil {
		return err
	}
	signalCh := workflow.GetSignalChannel(ctx, "my-signal")
	for {
		signalCh.Receive(ctx, nil)
		signalCount++
	}
}

func bToMb(b uint64) uint64 {
	return b / 1024 / 1024
}

```

</details>

I commented the line 107, pls test it first with that line commented and then uncomment it and test again

```go
	// _ = c.SignalWorkflow(ctx, run.GetID(), "", "my-signal", "some-arg")
```

With the line commented, you should see that after executing the last query the memory usage is like 

```
Alloc = 12 MiB  TotalAlloc = 1776 MiB   Sys = 35 MiB    NumGC = 314
```
You can also check the complete log to see how it is increased with every query.

With the line uncommented, you should see something like

```
Alloc = 4 MiB   TotalAlloc = 51 MiB     Sys = 19 MiB    NumGC = 23
```

The second test I think shows what I'm seeing, after a signal the cache is set, and from there the queries reuse it.

Also notice that executing the same 500 queries is faster.

Hope this helps.

**Comment by cretz** [comment]
Thanks! I will check and confirm whether we cache query-only tasks. It is quite possible that we do not because it is possible to query completed workflows and we have intentionally chosen not to fill the cache for just queries which don't actually affect workflow execution. This only becomes obvious on large history size.

I will do some investigation and confirm and see what options we have here.

**Comment by cretz** [comment]
I have opened a PR at #779 to fix the goroutine/mem leak that is occurring from queries over and over. If you add this to your `go.mod`:

    replace go.temporal.io/sdk => github.com/cretz/temporal-sdk-go query-goroutine-leak

And then run `go mod tidy` to turn that into a proper pseudo-version and then run the code again, you will see the memory usage now does not grow on repeated query invocations.

As for whether query-only workflow tasks should reuse a cache or not, that's a bit of a larger question and if there's a good enough use case we can look into make that opt-in.

**Comment by ovargasmahisoft** [comment]
@cretz Awesome, thanks a lot.

**Comment by cretz** [comment]
This has been merged and should be released in a couple weeks. There should not be a leak anymore in this use case. I am closing the issue, but feel free to comment/reopen if it persists.

If the need to reuse a cache for query-only tasks is really necessary, which I am not sure it is, we can discuss in Slack or a new issue.
